[#3] Types and scaffolding

lierdakil opened the pull request:
## Description

Problem: With the contract specification approved it is now possible to
define the types involved in the contract, Parameter and Storage.
It would be also useful to define the typical project structure and
scaffolding.

Solution: Define Parameter and Storage, add stub implementations
for entrypoints, add stub CLI.

***NB:*** I would prefer to split `Parameter` into admin/voting entrypoints types, but after fiddling with this for far too long, I gave up. Long story short, autodoc output ends up suboptimal at best and broken at worst. Perhaps I'm just missing something.

***UPD:*** I initially thought we could use `KeyHash` for admin and maintainers, but upon some reflection, it seems like a somewhat pointless artificial limitation, so I'm going with `Address` instead in dd63f13

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->
Resolves #3 

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

Barely anything to test yet.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)


- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

pasqu4le commented on src/Indigo/Contracts/HomebaseLite/Types.hs:
```diff
@@ -0,0 +1,67 @@
+
+data Storage = Storage
+  { sAdmin :: KeyHash
+  , sAdminCandidate :: Maybe KeyHash
```
A common trick for efficiency here (that we used in other contracts) is to store this without the `Maybe`.

The case in which there is no candidate is stored as if the candidate is the current `admin`.

This reduces the size of the stored data (admittedly very little), but also simplifies the implementation, because it removes the need to wrap/unwrap the `address` as well as some other check that would need to happen, while still fitting in the logic described in the spec.

pasqu4le commented on src/Indigo/Contracts/HomebaseLite/Types.hs:
```diff
@@ -0,0 +1,67 @@
+[entrypointDoc| Parameter plain |]
+
+data Configuration = Configuration
+  { cExpireTime :: Natural
+  , cVoteDelay :: Natural
```
What do you think about making a newtype for `Seconds` for clarity?

BTW `nat`s can't be subtracted from `timestamp`s in Michelson, so a conversion to `int` will be needed, however I still think it makes sense to pick `nat`s here.

pasqu4le commented on src/Indigo/Contracts/HomebaseLite/Types.hs:
```diff
@@ -0,0 +1,67 @@
+  | Remove_maintainers [KeyHash]
+  | Configure Configuration
+  | Propose (URI, [MText])
+  | Vote (URI, Natural)
```
Arguably a newtype for this `Natural` here would be useful as well, given that at a first glance it's not really clear that this denotes a choice.

Alternatively (I actually recommend this), you could make a used `Named` here for each side of the tuple (as well as for the argument of `Propose`) and make a type synonym as well for convenience.

pasqu4le commented on src/Indigo/Contracts/HomebaseLite/Impl/Vote.hs:
```diff
@@ -0,0 +1,50 @@
+    Called by a `participant` to submit a new proposal.
+    Takes as input an IPFS URI and a list of strings for the `n`  available choices.
+
+    - fails if the sender does not have a governance token amount of `minimum_balance` or more
+    - fails if a proposal for the same IPFS URI was already submitted before
+    - fails if the number of choices `n` is `0`
```
I'm mentioning this here, but it applies to everywhere else too, since errors have docs, it should be unnecessary to explain these here and instead the error documentation should be:
- defined with the `errorDoc` quasiquote (as you already do)
- included automatically in the produced documentation both in each entrypoint that uses them as well as in a dedicated section

lierdakil replied:
FWIW, I found this pretty useful while implementing, so there's that.

pasqu4le replied:
:thinking: I see, maybe I was too eager to suggest this and they could have been removed later on

pasqu4le commented on src/Indigo/Contracts/HomebaseLite/Impl/Admin.hs:
```diff
@@ -0,0 +1,76 @@
+  pass
+
+
+acceptAdmin :: tp :~> () => IndigoEntrypoint tp
```
Oh boy, I forgot how uncomfortable these signatures are to use and look at, I'll take a note to improve Indigo on it.

lierdakil replied:
FWIW, I've defined a synonym for this unpleasantness in #8, which makes it slightly less horrible.

pasqu4le commented on src/Indigo/Contracts/HomebaseLite/Types.hs:
```diff
@@ -0,0 +1,67 @@
+  = Set_admin KeyHash
+  | Accept_admin ()
+  | Add_maintainers [KeyHash]
+  | Remove_maintainers [KeyHash]
```
:thinking: I remember now that we discussed about making these Sets and that I provided some explanation for why lists are more often used.

However I'm not particularly against set in this case and I don't think that the `list` in the spec should be taken too literally (or it could be updated).

What do you think?

lierdakil replied:
At the end of the day, it doesn't really matter for the implementation, so... I don't know, it doesn't seem like there's much point to using a set.

pasqu4le replied:
Yeah, the only advantage is the one you noted of not permitting duplicates in the parameter with sets.

However admittedly this doesn't change the logic (even if there are duplicates the result will be the same) and it will only add a cost for those that are not careful enough.

Either option sounds ok to me as well :shrug:

pasqu4le commented on app/Main.hs:
```diff
@@ -1,9 +1,85 @@
+          Opt.showDefault
+    expireTimeParser = naturalParser #expireTime (7 * 24 * 3600)
+      "After how much time a proposal can no longer be voted on (in seconds)"
+    voteDelayParser = naturalParser #voteDelay (24 * 3600)
+      "After how much time votes can be cast on a proposal (in seconds)"
+    quorumThresholdParser = naturalParser #quorumThreshold 1
+      "The minimum amount of votes to be cast on a proposal for it to be successful, \
+      \expressed as a constant number to reach"
+    minimumBalanceParser = naturalParser #minimumBalance 1
+      "The minimum amount of governance tokens needed to submit a new proposal"
```
:thinking: arguably `0` is a possible (ans sensible, if that's what one wants) value for all of these, so how about using that as the default?

lierdakil replied:
`quorumThreshold` of `0` doesn't really make sense AFAIU -- that would imply all proposals are automatically successful regardless of votes.

pasqu4le replied:
well, yes, all proposals would be successful, but this wouldn't make them useless as there would still be the data of the choices made by those that participated (if any).

For example one may want to collect feedback from a community regardless of participation in the voting itself.

Arguably a bit of a stretch, but I think that "always successful proposals" are a valid option :shrug:

lierdakil replied:
Okay, fair.

pasqu4le commented on docs/documentation.md:
```diff
@@ -0,0 +1,470 @@
+
+
+
+
```
One more comment about this is that we generally add pipeline steps to generate and publish this (see [tezos-btc's example](https://github.com/tz-wrapped/tezos-btc/blob/master/.buildkite/pipeline.yml#L93-L140), so including the result in the repo is unnecessary.

I think this a dedicated "publishing" issue is better suited for most of those steps (including non-doc-related actions, e.g. for a static binary), but if you want to add the most basic autodoc step that runs on every PR here that would be good.

pasqu4le replied:
BTW at support of this comment, you didn't update it after the `KeyHash`->`Address` change :slightly_smiling_face:

pasqu4le commented on src/Indigo/Contracts/HomebaseLite/Types.hs:
```diff
@@ -13,10 +13,10 @@ module Indigo.Contracts.HomebaseLite.Types
 
 data Parameter
-  = Set_admin KeyHash
+  = Set_admin Address
```
:+1: on this change BTW

pasqu4le reviewed (commented):
I have some comments, but great work!

It looks like you didn't really miss anything of importance :smile:

lierdakil commented on .gitignore:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: LicenseRef-MIT-TC
 
-.stack-work/
+/.stack-work
```
I tend to symlink `.stack-work` to `tmpfs`, otherwise I can brew myself a cup of coffee _and_ drink it while the executable is linking :sweat_smile: Trailing `/` makes it so symlink is considered a valid target by git, which is a bit annoying.

lierdakil commented on src/Indigo/Contracts/HomebaseLite.hs:
```diff
@@ -2,8 +2,16 @@
 module Indigo.Contracts.HomebaseLite
-  ( todo
-  ) where
+    ( Parameter(..)
```
>  why the 4-space indentation here instead of the usual 2?

Probably a slip of the finger.

lierdakil commented on docs/documentation.md:
```diff
@@ -0,0 +1,470 @@
+
+# Homebase Lite
+
+
```
So... with the contract code for a `recursive` `Parameter` definition, I figured there are two options, either go with

```haskell
case_ param 
  ( #sAdmin #= adminEntrypoints
  , #sVoting #= votingEntrypoints
  )
```

In this case, autodoc says to wrap the parameter appropriately and call the default entrypoint. Not ideal, to say the least.

The other option is to use `entryCase`:


```haskell
entryCase (Proxy @PlainEntrypointsKind) param 
  ( #sAdmin #= adminEntrypoints
  , #sVoting #= votingEntrypoints
  )
```

But then `Admin` and `Voting` constructors are treated as entrypoints in their own right by autodoc. Also not ideal.

pasqu4le replied:
I see, it took me a while to look into this, but I think it makes sense to keep the current choice.

Apparently this can be fixed, by using `delegate` for the outer `Parameter` (and `plain` for the inner ones) with `entryCase`, _however_ you cannot use `PlainEntrypointsKind` for both levels (or you'll see the issue that you described) because these are at the same "level" AFAIU.
The solution would be to define and use some other `*EntrypointKind`, for an example see `morley-ledgers` and `FA2EntrypointsKind`+`FA2SampleParameter` and its [autodoc result](https://gitlab.com/morley-framework/morley-ledgers/-/blob/autodoc/master/autodoc/FA2Sample.md#section-FA2EntrypointsKind).

However looking at that, I'd say that it's also not ideal (due to the TOC) even tho the "how to call" sections are just right.

So, I suggest keeping the current choice of having a single `Parameter` and I think this is another thing worthy of re-consideration in `morley` itself :thinking:

lierdakil commented:
I believe I've addressed all comments, except switching to `set` (not necessarily sure it's a good idea), and I've kept the default `quorumThreshold` at `1`, for reasons explained above.

pasqu4le reviewed (approved):
LGTM now :ok_hand: 

I suppose that this now clashes a bit with #10 and that an entrypoint is missing, buy I'm ok with adding it in whichever one of them get merged last

lierdakil commented:
@pasqu4le, please take a quick look at https://github.com/tezos-commons/homebase-lite/pull/7/commits/fb715605c8319cb0c4b6265228d2ac7bc190721f and give the go-ahead.

pasqu4le commented:
@lierdakil checked, LGTM :+1:
